modules/ismp/clients/tendermint/verifier/src/verifier.rs

```diff
@@ -0,0 +1,540 @@
+use std::time::Duration;
+
+use tendermint::{block::Height, chain::Id, Hash, Time};
+use tendermint_light_client_verifier::{
+	options::Options,
+	types::{TrustThreshold, TrustedBlockState, UntrustedBlockState, ValidatorSet},
+	ProdVerifier, Verdict, Verifier,
+};
+use tendermint_proto::google::protobuf::Timestamp;
+
+use crate::{
+	ConsensusProof, TrustedState, UpdatedTrustedState, VerificationError, VerificationOptions,
+};
+
+/// Main verification function for header updates
+pub fn verify_header_update(
+	trusted_state: TrustedState,
+	consensus_proof: ConsensusProof,
+	options: VerificationOptions,
+	current_time: u64,
+) -> Result<UpdatedTrustedState, VerificationError> {
+	if trusted_state.is_frozen() {
+		return Err(VerificationError::Invalid("Trusted state is frozen".to_string()));
+	}
+
+	let chain_id = Id::try_from(trusted_state.chain_id.clone())
+		.map_err(|e| VerificationError::Invalid(e.to_string()))?;
+	let height = Height::try_from(trusted_state.height)
+		.map_err(|e| VerificationError::Invalid(e.to_string()))?;
+	let timestamp = Timestamp { seconds: trusted_state.timestamp as i64, nanos: 0 };
+	let time = Time::try_from(timestamp).map_err(|e| VerificationError::Invalid(e.to_string()))?;
+	let next_validators = ValidatorSet::new(trusted_state.next_validators.clone(), None);
+	let next_validators_hash = Hash::Sha256(trusted_state.next_validators_hash);
+
+	let tendermint_trusted_state = TrustedBlockState {
+		chain_id: &chain_id,
+		header_time: time,
+		height,
+		next_validators: &next_validators,
+		next_validators_hash,
+	};
+
+	let validators = ValidatorSet::new(consensus_proof.validators.clone(), None);
```

Comment:
We should use the validators from the trusted state
```suggestion
	let validators = ValidatorSet::new(consensus_proof.validators.clone(), None);
```